IPython/core/completer.py

```diff
@@ -2486,12 +2486,22 @@ def python_func_kw_matches(self, text):
         return argMatches
 
     @staticmethod
-    def _get_keys(obj: Any) -> List[Any]:
+    def _get_keys(obj: Any, prefix: Optional[str] = None) -> List[Any]:
         # Objects can define their own completions by defining an
         # _ipy_key_completions_() method.
-        method = get_real_method(obj, '_ipython_key_completions_')
-        if method is not None:
-            return method()
+        key_completions_getter = get_real_method(obj, "_ipython_key_completions_")
+
+        if key_completions_getter is not None:
+            # older versions of ipython assumed _ipython_key_completions_ took no arguments
+            if "prefix" in inspect.signature(key_completions_getter).parameters:
+                if prefix is not None:
+                    # strip leading apostrophe from key before passing to ._ipython_key_completions
+                    return key_completions_getter(prefix=prefix[1:])
+                else:
+                    return key_completions_getter(prefix=None)
+
```

Comment:
What I was suggesting was that instead of passing a `prefix` string we would pass an entire [`CompletionContext`](https://ipython.readthedocs.io/en/stable/api/generated/IPython.core.completer.html#IPython.core.completer.CompletionContext), something like:

```python
class ItemCompletionContext(CompletionContext):
    """Completion context for item sub-matchers (providing information about `__getitem__`)"""
    prefix: str
```

```suggestion
            if (
                 hasattr(key_completions_getter, 'matcher_api_version')
                 and key_completions_getter.matcher_api_version == 2
             ):
                context = ItemCompletionContext(
                    key_prefix=key_prefix,
                    full_text=context.full_text,
                    cursor_line=context.cursor_line,
                    cursor_position=context.cursor_position,
                    token=context.token,
               )
                return key_completions_getter(context)['completions']
```

Reply:
As for stripping leading apostrophe: what if the key is a number? Or hex-number? Or raw string? I think we do not want to modify the prefix.